java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ConfigValueCache.java

```diff
@@ -0,0 +1,296 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.lsp.server.protocol;
+
+import com.google.gson.JsonElement;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import org.eclipse.lsp4j.ConfigurationItem;
+import org.eclipse.lsp4j.ConfigurationParams;
+
+/**
+ * Cache for configuration values with support for hierarchical keys and change
+ * notifications
+ *
+ * @author atalati
+ */
+public class ConfigValueCache {
+
+    private final ConcurrentHashMap<String, Object> rootCache = new ConcurrentHashMap<>();
+    private static final String ROOT_KEY_VALUE = "";
+
+    public void registerListener(String section, BiConsumer<String, JsonElement> listener) {
+        ConfigData configData = getCachedConfigData(section);
+        if (configData != null) {
+            configData.setConsumer(listener);
+        } else {
+            setConfigInTree(section, new ConfigData(listener));
+        }
+    }
+
+    public void registerCache(String section) {
+        ConfigData configData = getCachedConfigData(section);
+        if (configData == null) {
+            setConfigInTree(section, new ConfigData());
+        }
+    }
+
+    public JsonElement getConfigValue(String section, String scope) {
+        ConfigData configData = getCachedConfigData(section);
+        if (configData == null) {
+            return null;
+        } else if (scope == null) {
+            return configData.getRootValue();
+        }
+        return configData.getScopedValue(scope);
+    }
+
+    public void updateCache(NbCodeLanguageClient client, String section, Object cacheValue, JsonElement tree) {
+        if (tree == null || cacheValue == null) {
+            return;
+        }
+
+        ConfigData rootData;
+        ConfigValueCache currentCache = null;
+
+        if (isConfigDataInstance(cacheValue)) {
+            rootData = (ConfigData) cacheValue;
+        } else if (isConfigValueInstance(cacheValue)) {
+            currentCache = (ConfigValueCache) cacheValue;
+            rootData = currentCache.getRootCacheValue();
+        } else {
+            return;
+        }
+
+        if (rootData != null) {
+            rootData.setRootValue(tree);
+            try {
+                BiConsumer<String, JsonElement> consumer = rootData.getConsumer();
+                if (consumer != null) {
+                    consumer.accept(section, tree);
+                }
+            } catch (RuntimeException e) {
```

Comment:
Catching and throwing away an exception like this is usually wrong. If there's something wrong, and the exception happens, we will never find out, things will just appear broken.

In cases there's a valid reason to continue after a callback fails with an exception, the exception should be logged, so that at least from the log we can find out what happened.

In cases we don't see a valid reason why the callback would fail, I would simply not catch the error.